design/src/rfcs/rfc0031_implementing_adaptive_retry_behavior.md

```diff
@@ -0,0 +1,162 @@
+
+This RFC defines how the Rust SDK will implement Adaptive Retry Behavior. This new behavior works on the principle of client-side rate limiting and only takes effect once a client has received a throttling error. The new behavior:
+- Is opt-in. The existing retry behavior will remain the default.
+- Will perform similarly to the existing retry behavior when comparing the rate of successful requests.
+- Will improve the experience of customers that use the SDK to send very large numbers of concurrent requests.
+
+*This RFC does not consider retry behavior related to expired credentials or clock skew errors.*
+
+## Terminology
+
+- **Retry (Behavior / Strategy)**: How the SDK reacts to retryable errors.
+- **Standard Retry Behavior (SRB)**: The default behavior. When a retryable error is received, this behavior calculates an exponential backoff time based on how many retry attemps have already been made and then sleeps for that long before making another attempt.
+- **Adaptive Retry Behavior (ARB)**: When sending requests, clients must acquire a token from a shared bucket. When responses are successful, the bucket capacity and fill rate are increased. When throttling responses are received, the bucket capacity and fill rate are decreased. **Transient Errors** do not affect the rate of token replenishment. When the bucket is empty, requests will sleep until more tokens become available.
+- **Token Bucket**: similar to a [`tokio::sync::Semaphore`][tokio::sync::Semaphore], a token bucket contains a limited amount of "tokens." Before making an HTTP request, we take a token from the bucket. If the request fails, we take another token and try again. If the request succeeds, we release any tokens we're holding. If the token bucket is empty, no requests may be made.
+- [**Retryable Trait**](https://awslabs.github.io/smithy/spec/core.html#retryable-trait): a smithy trait indicating that an error may be retried by the client.
+- **(HTTP / Retry) Request**: An HTTP request. One or more of these will be made during an **Operation Request**.
+- **Operation Request**: A request to a smithy service. Each operation request may contain several **HTTP requests**.
+- **Throttling Error**: An error that should trigger the client's rate-limiting implementation. Currently, the Rust SDK has no special behavior for these errors. Once this RFC is implemented, the Rust SDK will treat the following as throttling errors:
+    - `400` errors with any of the following codes:
+        - `Throttling`
+        - `ThrottlingException`
+        - `ThrottledException`
+        - `RequestThrottledException`
+        - `TooManyRequestsException`
+        - `ProvisionedThroughputExceededException`
+        - `TransactionInProgressException`
+        - `LimitExceededException`
+        - `PriorRequestNotComplete`
+    - `403` errors with the `RequestThrottled` code.
+    - `502` errors with the `EC2ThrottledException` code.
```

Comment:
Just to clarify, a 502 that does not have `EC2ThrottledException` is still going to be a transient error?

Also, are the codes case sensitive?